classes/ActionScheduler_Data.php

```diff
@@ -0,0 +1,185 @@
+<?php
+
+use Action_Scheduler\Migration\ActionScheduler_MigrationConfig;
+use Action_Scheduler\Migration\ActionScheduler_MigrationScheduler;
+use Action_Scheduler\WP_CLI\ActionScheduler_WPCLI_Migration_Command;
+use Action_Scheduler\WP_CLI\ActionScheduler_WPCLI_ProgressBar;
+
+/**
+ * Class ActionScheduler_Data
+ *
+ * The main plugin/initialization class for the
+ * Action Scheduler Custom Tables plugin.
+ *
+ * Responsible for hooking everything up with WordPress.
+ *
+ * @package Action_Scheduler
+ *
+ * @codeCoverageIgnore
+ */
+class ActionScheduler_Data {
+	private static $instance;
+
+	/** @var ActionScheduler_MigrationScheduler */
+	private $migration_scheduler;
+
+	/** @var string */
+	private $store_classname;
+
+	/** @var string */
+	private $logger_classname;
+
+	/** @var bool */
+	private $migrate_custom_store;
+
+	/**
+	 * ActionScheduler_Data constructor.
+	 *
+	 * @param Migration_Scheduler $migration_scheduler
+	 */
+	public function __construct( ActionScheduler_MigrationScheduler $migration_scheduler ) {
+		$this->migration_scheduler = $migration_scheduler;
+		$this->store_classname     = '';
+	}
+
+	/**
+	 * Get the action store class name.
+	 *
+	 * @param string $class Classname of the store class.
+	 *
+	 * @return string
+	 */
+	public function get_store_class( $class ) {
+		if ( $this->migration_scheduler->is_migration_complete() ) {
+			return 'ActionScheduler_DBStore';
+		} elseif ( ActionScheduler_Store::DEFAULT_CLASS !== $class ) {
+			$this->store_classname = $class;
+			return $class;
+		} else {
+			return 'ActionScheduler_HybridStore';
+		}
+	}
+
+	/**
+	 * Get the action logger class name.
+	 *
+	 * @param string $class Classname of the logger class.
+	 *
+	 * @return string
+	 */
+	public function get_logger_class( $class ) {
+		ActionScheduler_Store::instance();
+
+		if ( $this->store_classname ) {
+			$this->logger_classname = $class;
+			return $class;
+		} else {
+			return 'ActionScheduler_DBLogger';
+		}
+	}
+
+	/**
+	 * Register the WP-CLI command to handle bulk migrations
+	 *
+	 * @return void
+	 */
+	public function register_cli_command() {
+		if ( defined( 'WP_CLI' ) && WP_CLI && $this->allow_custom_migration() && ! $this->migration_scheduler->is_migration_complete() && $this->migration_scheduler->dependencies_met() ) {
+			$command = new ActionScheduler_WPCLI_Migration_Command();
+			$command->register();
```

Comment:
Could we unify the registration of this command with the registration of the `ActionScheduler_WPCLI_Scheduler_command` command rather than having two separate CLI systems.

Reply:
> This confusion highlights that the logic in this method may be better moved to ActionScheduler_Store::instance() and this method can be deleted

AS is compatible with PHP 5.2 but the migration to CT won't kick in unless PHP is >= 5.6. I opted for adding filters in the CT code over adding 5.6 conditional checks in the main plugin code. That can be revisited. I'll look at better filter names to start.

Reply:
> Could we unify the registration of this command

The reason I added it that way is the command should only be registered if migration conditions are met and the migration isn't complete. The idea being that once the migration is done, none of the migration code is loaded.

Reply:
> the migration to CT won't kick in unless PHP is >= 5.6

Ah, I missed that critical part. That makes sense, and is a really good approach. While WP 5.2 & WC 3.7 require PHP 5.6, I like the idea of maintaining backward compatibility with 5.2 for a while longer in AS for older users.